### DISCUSSION TITLE
Indent whole line instead of adding a tab at caret position

### DISCUSSION THREAD

**Discussion by Cutipus** [godotengine/godot-proposals#6775] [Editor]
I would like "ui_text_indent" to indent a whole line, rather than adding a tab at caret position.

While coding I noticed that pressing "Tab" ("ui_text_indent") will insert a tab character at the caret's position. However, pressing "Shift+Tab" ("ui_text_dedent") will remove a tab character from the beginning of the line. This is confusing and inconsistent, as in order to indent a whole line, one would have to highlight a portion of the text and then press tab - i.e "Shift+Right" and then "Tab".

I would like to be able to indent a whole line without having to select parts of the text first, regardless of caret's position. It makes little sense to add tabs in the middle of a line, and the only use cases I can think of where indentation would be useful anywhere but the beginning of a line is when trying to align comments at the end of a line, or when trying to align key-value pairs in a dict definition.

`ui_text_indent` calls `do_indent()`, which is hard-coded to only call `indent_lines()` when the caret `has_selection`, as shown in [code_edit.cpp#L858](https://github.com/godotengine/godot/blob/9f12e7b52d944281a39b7d3a33de6700c76cc23a/scene/gui/code_edit.cpp#L858). It would be nice if a change were made to make it possible to indent the current line without having to select anything, or even making it the default action for `ui_text_indent` considering it is a much much more common use-case.

**Comment by Calinou** [comment]
> While coding I noticed that pressing "Tab" ("ui_text_indent") will insert a tab character at the caret's position. However, pressing "Shift+Tab" ("ui_text_dedent") will remove a tab character from the beginning of the line. This is confusing and inconsistent, as in order to indent a whole line, one would have to highlight a portion of the text and then press tab - i.e "Shift+Right" and then "Tab".

This is how most editors out there behave, such as Visual Studio Code. Dedent always works regardless of cursor option, but indent inserts a tab at cursor location as you may be using tabs for alignment (or writing literal tab characters in strings).

**Reply by Cutipus** [reply]
Visual Studio Code also has a keybind for indenting and outdenting that works as I have explained. "Ctrl+]" according to my search. Indenting means moving the whole line - whether or not the Tab key should default to that action is a different matter, I would just like the option to exist.

**Reply by Calinou** [reply]
I guess it makes sense to implement it as its own shortcut, but the default Tab action should remain as-is.
